database/migrations/default/1747404534187_patch_crash_ems_trigger/up.sql

```diff
@@ -0,0 +1,118 @@
+                    AND matching_ems.incident_received_datetime  <= (c.crash_timestamp + time_threshold)
+                    AND c.position IS NOT NULL
+                    AND ST_DWithin(matching_ems.geometry::geography, c.position::geography, meters_threshold)
+            ) INTO matched_crash_ids;
+            
+        -- Get the count from the array length (handling when array_length is null as 0)
+        SELECT COALESCE(array_length(matched_crash_ids, 1), 0) INTO match_count;
+
+        -- For records that have been manually matched we want to only update the matched_crash_pks column
+        IF matching_ems.crash_match_status = 'matched_by_manual_qa' THEN
+            UPDATE ems__incidents 
+            SET matched_crash_pks = matched_crash_ids
+            WHERE id = matching_ems.id;
+        -- If the record has not been manually matched
+        ELSE
+          IF match_count = 0 THEN
+              UPDATE ems__incidents 
+              SET crash_pk = NULL,
+                  crash_match_status = 'unmatched',
+                  matched_crash_pks = NULL
+              WHERE id = matching_ems.id;
+          ELSIF match_count = 1 THEN
+              -- this EMS record is only matched to one crash - we can assign the crash_pk and matched_crash_pks
+              UPDATE ems__incidents 
+              SET crash_pk = NEW.id,
+                  crash_match_status = 'matched_by_automation',
+                  matched_crash_pks = matched_crash_ids
+              WHERE id = matching_ems.id;
+          ELSE
+              -- multiple matching crashes found - can assign the matched_crash_pks but not crash_pk
+              UPDATE ems__incidents 
+              SET crash_match_status = 'multiple_matches_by_automation',
+                  crash_pk = NULL,
+                  matched_crash_pks = matched_crash_ids
+              WHERE id = matching_ems.id;
+          END IF;
+        END IF;
+    END LOOP;
+
+    IF TG_OP = 'UPDATE' THEN
+        -- update incidents which have not been manually matched
+        UPDATE ems__incidents
+        SET crash_pk = NULL,
+            crash_match_status = 'unmatched',
+            matched_crash_pks = NULL
+        WHERE crash_pk = NEW.id
+          AND (
+              incident_received_datetime  < (NEW.crash_timestamp - time_threshold)
+              OR incident_received_datetime  > (NEW.crash_timestamp + time_threshold)
+              OR geometry IS NULL
+              OR NEW.position IS NULL
+              OR NEW.is_temp_record is TRUE
+              or NEW.is_deleted is TRUE
+              OR (NOT ST_DWithin(geometry::geography, NEW.position::geography, meters_threshold))
+          )
+          AND 
+            -- ignore manual qa matches
+            crash_match_status != 'matched_by_manual_qa';
+       
+        -- update the matched_crash_pks column only for manually qa matches
+        UPDATE ems__incidents
+        SET matched_crash_pks = NULL
```

Comment:
@johnclary I follow this whole function, but I have one question about setting `matched_crash_pks` to `null` here but not in the block on lines 50-53. I think that I'm getting lost on when we want to use `matched_crash_pks` as a holding place for recovering previously matches or when we don't want to.

Reply:
i believe the logic here is that `matched_crash_pks` should always stay up to date with what our automated matching finds, and in this case this incident doesnt have any matches, so `matched_crash_pks` would update to null. but this wouldnt affect the manual value in `crash_pk` that the VZ team manually matched. in lines 50-53 the record did have matching crash ids, so thats why `matched_crash_pks` was updated there.

Reply:
ohhh now i see it. the WHERE conditions in this update statement are how we know there are no automated matches - that is what I was missing. 🙏 Thanks for helping me through this because I was looking earlier in the code and not seeing that kind of check.

Reply:
> when we want to use `matched_crash_pks as a holding` place for recovering previous matches and when we don't want to.

@mddilley Our intention is that `matched_crash_pks` _always_ holds the automatically matched crash PKs regardless of any user edits. This way we can reset to the original/automatic state of the EMS record. So the trigger is keeping this column in sync with any edits that are happening to the crash.

Reply:
I added this thread to https://github.com/cityofaustin/atd-data-tech/issues/22006